src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,497 @@
+
+        -- All the remainingUtxos that has not been used as an input to the transaction yet.
+        remainingUtxos :: [(TxOutRef, TxOut)]
+        remainingUtxos =
+          Map.toList $
+            Map.filterWithKey
+              (\k v -> k `notElem` txInRefs && isRight (txOutToTxIn (k, v)))
+              utxosIndex
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Remaining UTxOs: " <+> pretty remainingUtxos
+
+    -- the input vector for the current transaction, this can be a zero vector when there are no
+    -- inputs the transaction.
+    txInsVec <-
+      hoistEither $
+        if Value.isZero txInsValue
+          then Right $ zeroVec (length allAssetClasses)
+          else valueToVec allAssetClasses txInsValue
+
+    -- the output vector of the current transaction, this is all the values of TxOut combined.
+    outVec <- hoistEither $ valueToVec allAssetClasses outValue
+
+    -- all the remainingUtxos converted to the vectors.
+    remainingUtxosVec <- hoistEither $ mapM (valueToVec allAssetClasses . txOutValue . snd) remainingUtxos
+
+    -- we use the default search strategy to get indexes of optimal utxos, these indexes are for the
+    -- remainingUtxos, as we are sampling utxos from that set.
+    selectedUtxosIdxs <-
+      newEitherT $
+        searchTxIns @w
+          defaultSearchStrategy
+          (isSufficient outVec)
+          outVec
+          txInsVec
+          remainingUtxosVec
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "" <+> "Selected UTxOs Index: " <+> pretty selectedUtxosIdxs
+
+    let -- These are the selected utxos that we get using `selectedUtxosIdxs`.
+        selectedUtxos :: [(TxOutRef, TxOut)]
+        selectedUtxos = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxos ^? ix idx) . folded
+
+        selectedVectors :: [ValueVector]
+        selectedVectors = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxosVec ^? ix idx) . folded
+
+    finalTxInputVector <- hoistEither $ foldM addVec txInsVec selectedVectors
+    unless (isSufficient outVec finalTxInputVector) $ throwError "Insufficient Funds"
+
+    selectedTxIns <- hoistEither $ mapM txOutToTxIn selectedUtxos
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Selected TxIns: " <+> pretty selectedTxIns
+
+    -- Now we add the selected utxos to originalTxIns present in the transaction previously.
+    pure $ originalTxIns <> Set.fromList selectedTxIns
+  where
+    -- This represents the condition when we can stop searching for utxos.
+    -- First condition is that the input vector must not be zero vector, i.e.
+    -- There must be at least some input to the transaction.
+    -- Second condition is that all the values of input vector must be greater than
+    -- or equal to the output vector.
+    isSufficient :: ValueVector -> ValueVector -> Bool
+    isSufficient outVec txInsVec =
+      Vec.all (== True) (Vec.zipWith (<=) outVec txInsVec)
+        && txInsVec /= zeroVec (length txInsVec)
```

Comment:
In what cases do we need the extra clause that the `txInsVec /= zeroVec`, unless `outVec == zeroVec` this should be implied as well

Reply:
`txInsVec /= zeroVec` implies that there's is not input to the Tx, which is not correct I think as a Tx must have some input to create output.